app/models/tree_resolver.rb

```diff
@@ -25,53 +22,78 @@ def enqueue_tree_resolution
   end
 
   def load_dependencies_tree
-    tree_data = Rails.cache.fetch cache_key, :expires_in => 1.day, race_condition_ttl: 2.minutes do
-      generate_dependency_tree
-    end
+    tree_data = Rails.cache.fetch(
+      cache_key,
+      expires_in: 1.day,
+      race_condition_ttl: 2.minutes,
+      &method(:generate_dependency_tree)
+    )
     @project_names = tree_data[:project_names]
     @license_names = tree_data[:license_names]
     @tree = tree_data[:tree]
   end
 
+  def project_names
+    @project_names.to_a
+  end
+
+  def license_names
+    @license_names.to_a
+  end
+
   private
 
-  def generate_dependency_tree
+  def generate_dependency_tree(_key)
     {
       tree: load_dependencies_for(@version, nil, @kind, 0),
-      project_names: project_names,
-      license_names: license_names
+      project_names: @project_names,
+      license_names: @license_names,
     }
   end
 
   def load_dependencies_for(version, dependency, kind, index)
-    if version
-      @license_names << version.project.try(:normalized_licenses)
-      kind = index.zero? ? kind : 'runtime'
-      dependencies = version.dependencies.kind(kind).includes(project: :versions).limit(100).order(:project_name)
-      {
-        version: version,
-        dependency: dependency,
-        requirements: dependency.try(:requirements),
-        normalized_licenses: version.project.try(:normalized_licenses),
-        dependencies: dependencies.map do |dep|
-          if dep.project && !@project_names.include?(dep.project_name)
-            @project_names << dep.project_name
-            index < 10 ? load_dependencies_for(dep.latest_resolvable_version(@date), dep, kind, index + 1) : ['MORE']
-          else
-            {
-              version: dep.latest_resolvable_version(@date),
-              requirements: dep.try(:requirements),
-              dependency: dep,
-              normalized_licenses: dep.project.try(:normalized_licenses),
-              dependencies: []
-            }
-          end
-        end
-      }
-    end
+    return unless version
+
+    append_license_names(version)
+    should_fetch = append_project_name(dependency)
+
+    return if index > MAX_TREE_DEPTH
```

Comment:
Should we include anything so that we know that the max depth has been exceeded and that the tree is thus truncated?

Reply:
We talked about it, but I figured YAGNI applied here. I can add in a flag if you think it's important.

Reply:
A log message is probably a fine super lightweight start. Then we can see if we're hitting it often

Reply:
Done

> Max tree depth 15 reached when loading dependencies for Rubygems/16 for parent Rubygems/root